pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -2086,14 +2100,64 @@ impl Timeline {
         }
         // end: move the above part to the loop body
 
+        let mut rewrote_delta_layers = Vec::new();
+        if let Some(compaction_key_range) = &job_desc.partial_key_range {
+            for layer in job_desc.rewrite_layers {
+                // For each of the layer to rewrite, generate a layer with the same LSN but less key content
+                // This download should be no-op b/c it's included in selected_layer and has been downloaded before
+                let resident_layer = layer.download_and_keep_resident().await?;
+                let desc = resident_layer.layer_desc();
+                let layer = resident_layer.get_as_delta(ctx).await?;
+                let mut iter = layer.iter(ctx);
+                let mut delta_writer_before = DeltaLayerWriter::new(
+                    self.conf,
+                    self.timeline_id,
+                    self.tenant_shard_id,
+                    desc.key_range.start,
+                    desc.lsn_range.clone(),
+                    ctx,
+                )
+                .await?;
+                let mut delta_writer_after = DeltaLayerWriter::new(
+                    self.conf,
+                    self.timeline_id,
+                    self.tenant_shard_id,
+                    compaction_key_range.end,
+                    desc.lsn_range.clone(),
+                    ctx,
+                )
+                .await?;
+                while let Some((key, lsn, val)) = iter.next().await? {
+                    if key < compaction_key_range.start {
+                        delta_writer_before.put_value(key, lsn, val, ctx).await?;
+                    } else if key >= compaction_key_range.end {
+                        delta_writer_after.put_value(key, lsn, val, ctx).await?;
+                    }
+                }
```

Comment:
We are re-reading the entire layer here, this is inefficient.

Why not put the logic to rewrite partially overlapping deltas into the k-merge?

Reply:
Like, create batch delta layer writers for before and after the partial range before we start the k-merge loop starts.
Feed data into these writers from this location:

https://github.com/neondatabase/neon/blob/24cd9101408df66f3ffa0497b936da50447785dd/pageserver/src/tenant/timeline/compaction.rs#L2039

Reply:
https://neondb.slack.com/archives/C06Q6MQ2QSC/p1730740354684789

Reply:
#9671